### PR TITLE
Bulk insert aggregated results

### PR DESCRIPTION
**Pull request by rafamanzo**
This is mostly @danielkza a great work for which I've split some commits into atomic ones.

To improve even more the bulk insertions effect, walking by levels bottom up has been added alongside this same PR.

As this stores a lot of data in memory, I think adding memory statistics to the profiling is sane before accepting this.

Aggregation performance results:

Aggregation performance test results below:

Metric Count | Branch |  Wall Time (s) | Process Time (s)
-----------------------|-------------|-------------------------|------------------------------
1 | v1.3.2 | 34.16732797622681 | 29.564079139199997
1 | optimize_aggregation_reviewed | 4.813459873199463 | 6.0212057172
2 | v1.3.2 | 71.57923197746277 | 63.06931912939999
2 | optimize_aggregation_reviewed | 7.97909255027771 | 9.9699096154
4 | v1.3.2 | 154.1164906024933 | 147.8050995992
4 | optimize_aggregation_reviewed | 14.562432861328125 | 17.795247482599997
8 | v1.3.2 | 333.3348692417145 | 311.43665788299995
8 | optimize_aggregation_reviewed | 33.7912938117981 | 35.3004300856

Almost a 10x speedup! Nicely done!

This is part of https://github.com/mezuro/kalibro_processor/issues/207.

After accepting this, please create an issue so we do not forget to rewrite `MetricResult#pre_order` unit tests to take full advantage of the context already existent objects.

### CODE REVIEW

**Comment by danielkza** [comment]
I've ran some tests and memory usage is actually better, probably due to not fetching multiple copies of the same entities repeatedly. [Gist](https://gist.github.com/danielkza/521a566c895a53ae99f11b012bb8d21a)

File: CHANGELOG.rdoc [hunk 1]
```diff
@@ -4,6 +4,9 @@ KalibroProcessor is the processing web service for Mezuro.
 
 == Unreleased
 
+* Insert in one query all aggregated MetricResults
```
Comment (diegoamc) [inline CHANGELOG.rdoc]: I don't mean to be too meticulous here, but now we are __batching__ the queries, right?

Comment (danielkza) [inline CHANGELOG.rdoc]: @diegoamc Right. I personally prefer changelog entries that describe changes as they would affect a user reading it, and not exactly what changed. In this case it doesn't really matter whether we batch, insert at once, or whatever else, only that the aggregation got faster.

File: app/models/module_result.rb [hunk 2]
```diff
@@ -30,8 +30,27 @@ def pre_order
     @pre_order ||= pre_order_traverse(root).to_a
   end
 
+  def level_order
```
Comment (diegoamc) [inline app/models/module_result.rb]: Is this method actually called somewhere?

Comment (danielkza) [inline app/models/module_result.rb]: Nice catch, It isn't used. it was part of an earlier version of the PR. It
might be useful for other processing steps, but we can add it whenever we
actually need it.

Em qua, 27 de jul de 2016 15:28, Diego de Araújo Martinez Camarinha <
notifications@github.com> escreveu:

**Comment by diegoamc** [comment]
Congratulations, this is a very fine piece of work! :clap:

I think we should remove `#level_order` along with its tests before merging. What do you think?

**Comment by danielkza** [comment]
Good to me!

Em qua, 27 de jul de 2016 15:39, Diego de Araújo Martinez Camarinha <
notifications@github.com> escreveu:

**Comment by danielkza** [comment]
Done.

**Comment by diegoamc** [comment]
Sweet :candy:

**Pull request merged**